### PR TITLE
Distinguish runtime comm consistency unordered-forall, task-end funcs.

### PR DESCRIPTION
**Pull request by gbtitus**
For some time we've had a comm layer function which ensures remote
memory consistency after a series of unordered comm operations, but
whose name (`chpl_comm_task_end()`) implies that it is related to task
termination.  There was even a TODO comment in `ChapelBase.chpl` saying
that at least one call to it should be changed to something implying a
relationship with unordered operations.  Here, do that.  Add a new
`chpl_comm_unordered_consist_fence()`.  Change `chpl_comm_task_end()` calls
following unordered foralls to go to `chpl_comm_unordered_consist_fence()`
instead.  Retain calls to `chpl_comm_task_end()` in `_downEndCount()`, thus
at task ending.  Implement `chpl_comm_unordered_consist_fence()` in the
two comm layers (ugni and ofi) that actually need it.

While here, also adjust the runtime declarations for these to use our
customary mechanism for optionally-implemented functions, and remove the
empty versions of `chpl_comm_task_end()` from comm=none and comm=gasnet,
since those don't actually need that function.

Signed-off-by: Greg Titus <gbtitus@users.noreply.github.com>

### CODE REVIEW

File: modules/internal/ChapelBase.chpl [hunk 1]
```diff
@@ -1264,17 +1264,18 @@ module ChapelBase {
...
   proc chpl_after_forall_fence() {
-    chpl_comm_task_end(); // TODO: change to chpl_comm_unordered_task_fence()
+    chpl_comm_unordered_consist_fence();
   }
 
   // This function is called once by each newly initiated task.  No on
   // statement is needed because the call to sub() will do a remote
   // fork (on) if needed.
+  pragma "task complete impl fn"
+  extern proc chpl_comm_task_end(): void;
```
Comment (mppf) [inline modules/internal/ChapelBase.chpl:1277]: The comment above is supposed to apply to downEndCount. Also, can this extern proc declaration be next to chpl_comm_unordered_consist_fence above?

File: modules/internal/ChapelBase.chpl [hunk 2]
```diff
@@ -1264,17 +1264,18 @@ module ChapelBase {
     }
   }
 
-  pragma "task complete impl fn"
-  extern proc chpl_comm_task_end(): void;
+  extern proc chpl_comm_unordered_consist_fence(): void;
```
Comment (mppf) [inline modules/internal/ChapelBase.chpl:1267]: It might need `  pragma "task complete impl fn"` for the forall unordered opt to work right. We can add another flag if that turns out to be the case.

**Review by mppf — APPROVED** [review]

**Comment by gbtitus** [comment]
Just a note about testing: I've completed a `start_test -perflabel ml-` run against this on XC/comm-ugni with all passing, and am currently running a `start_test` on CS of all the directories (non-recursive) which contain a `.chpl` file with "unorder" (case insensitive)  in it.  I also did minimal testing with comm=none and comm=gasnet, since for those the change was just a removed definition in the runtime.

File: modules/internal/ChapelBase.chpl (hunk 1 above)
Comment (gbtitus) [reply modules/internal/ChapelBase.chpl:1277]: Sure.  I was undecided whether the `extern` decls should be right before their first use or just somewhere before.

File: compiler/optimizations/optimizeForallUnorderedOps.cpp [hunk 3]
```diff
@@ -358,7 +358,7 @@ void checkLifetimesForForallUnorderedOps(FnSymbol* fn,
 
   // This runs even for fNoOptimizeForallUnordered
   // because even if the optimization is disabled, we want
-  // to mark the ends of foralls with chpl_comm_unordered_task_fence.
+  // to mark the ends of foralls with chpl_comm_unordered_consist_fence.
```
Comment (ronawho) [inline compiler/optimizations/optimizeForallUnorderedOps.cpp:361]: There's also a chpl_gen_comm_getput_unordered_task_fence

File: modules/internal/ChapelBase.chpl (hunk 2 above)
Comment (gbtitus) [reply modules/internal/ChapelBase.chpl:1267]: I looked at how the compiler used this and the best I could tell, the `pragma "task complete impl fn"` actually did seem to have to do with task termination rather than unordered operations.  Testing seems to indicate (so far) that unordered ops are getting completed and made consistent properly.

File: compiler/optimizations/optimizeForallUnorderedOps.cpp (hunk 3 above)
Comment (ronawho) [reply compiler/optimizations/optimizeForallUnorderedOps.cpp:361]: And a chpl_comm_atomic_unordered_task_fence. I'm not sure either needs to change as part of this, but might be worth it for consistency. 

I'd also be fine with changing this back to chpl_comm_unordered_task_fence. I'm not sure consist is all that more meaningful to me. (The "task" in the name implies that it only fences operations initiated by this task, it's not denoting a task end)

Comment (gbtitus) [reply compiler/optimizations/optimizeForallUnorderedOps.cpp:361]: Or actually `chpl_comm_getput_unordered_task_fence()`?  (Not `chpl_gen_comm_*`.)  I'll do the `_IMPL` thing with that one also.  I didn't even notice it.

Comment (gbtitus) [reply compiler/optimizations/optimizeForallUnorderedOps.cpp:361]: And there is also `chpl_comm_atomic_unordered_task_fence()`.  I'm happy to do that one also.  In the interest of consistency should I name them all the same?  `_task_fence` or `_consist_fence`?  Should I put them together in the source files or leave them apart (as they are now)?  Could this additional work wait?  If I can get this PR in today then I can maybe get my comm=ofi MCM conformance improvements in for the release, which would make Shasta support a bit simpler.

Comment (ronawho) [reply compiler/optimizations/optimizeForallUnorderedOps.cpp:361]: There's `chpl_gen_comm_getput_unordered_task_fence` and `chpl_cache_comm_getput_unordered_task_fence` and I think `chpl_comm_getput_unordered_task_fence`. There's a bunch of these. 

There's the get/put ones in the runtime, the atomic ones in the runtime, and then this higher level unordered one that fences both puts/gets and atomics. 

What's the motivation for changing this name from `_task_fence` to `_consist_fence`? I'd probably just leave it as `_task_fence` so there's not so many names to change.

Comment (ronawho) [reply compiler/optimizations/optimizeForallUnorderedOps.cpp:361]: Oh, maybe i'm misreading this (got confused reading the comments here vs the module code change)  -- You changed `chpl_comm_task_end` to `chpl_comm_unordered_consist_fence`. Maybe just call that `chpl_comm_unordered_task_fence`?

Comment (gbtitus) [reply compiler/optimizations/optimizeForallUnorderedOps.cpp:361]: No, I added `chpl_comm_unordered_consist_fence()` and changed some calls that referenced `chpl_comm_task_end()` to call that new one instead.  `chpl_comm_task_end()` still exists, and is still called from other places.  I'll change `chpl_comm_unordered_consist_fence()` to `chpl_comm_unordered_task_fence()`.

**Comment by gbtitus** [comment]
I had to restart the Cray CS testing, so at this point I'm leaning toward delaying this until Monday 9/28 morning due to my intervening away time.

**Review by ronawho — APPROVED** [review]

**Comment by ronawho** [comment]
> I had to restart the Cray CS testing, so at this point I'm leaning toward delaying this until Monday 9/28 morning due to my intervening away time.

Sorry, I missed the note about trying to get this in today and then had to run to a meeting.

**Comment by gbtitus** [comment]
It turned out this failed a few tests in Cray CS comm=ofi testing, so I'll need to figure out what's going on and repair as necessary before it can be merged.

**Comment by gbtitus** [comment]
@mppf Would you take a quick look at the little change I just committed?  With this all the tests containing "unord" are passing on a CS system with comm=ofi and the verbs provider.

**Review by mppf — APPROVED** [review]
@gbtitus - I think this is OK for now but it would make more sense to me for `pragma "compiler added remote fence"` to be on the function `chpl_after_forall_fence` rather than on the `extern proc chpl_comm_unordered_task_fence`. It seems that there are no other calls to `chpl_comm_unordered_task_fence` today and so as a result these two should work the same. However if we end up in a situation where `chpl_comm_unordered_task_fence` needs to be called and it is not called by the compiler, I wouldn't want that flag to apply to it.

**Comment by gbtitus** [comment]
> ... OK for now but it would make more sense ...

Oh, you're so right!  I'll go fix this.  It'll take almost no time to re-test just the failures after doing so.  Thanks!
